### PR TITLE
K8SPXC-851: handle replication password change

### PR DESCRIPTION
**Pull request by qJkee**
[K8SPXC-851](https://jira.percona.com/browse/K8SPXC-851) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

Update password for replication channel if replication password changed

### CODE REVIEW

File: deploy/cr.yaml [hunk 1]
```diff
@@ -36,10 +36,10 @@ spec:
...
     image: percona/percona-xtradb-cluster:8.0.23-14.1
     autoRecovery: true
-#    expose:
-#      enabled: true
-#      type: LoadBalancer
-#      trafficPolicy: Local
+#     expose:
+#       enabled: true
+#       type: LoadBalancer
+#       trafficPolicy: Cluster
```
Comment (egegunes) [inline deploy/cr.yaml:42]: why are we changing the `trafficPolicy`?

File: pkg/controller/pxc/users.go [hunk 2]
```diff
@@ -223,9 +223,11 @@ func (r *ReconcilePerconaXtraDBCluster) manageXtrabackupUser(cr *api.PerconaXtra
 	return nil
 }
 
-func (r *ReconcilePerconaXtraDBCluster) manageSysUsers(cr *api.PerconaXtraDBCluster, sysUsersSecretObj, internalSysSecretObj *corev1.Secret) (bool, bool, error) {
+func (r *ReconcilePerconaXtraDBCluster) manageSysUsers(cr *api.PerconaXtraDBCluster, sysUsersSecretObj, internalSysSecretObj *corev1.Secret) (bool, bool, bool, error) {
 	type action int
 
+	replicaPassChanged := false
```
Comment (egegunes) [inline pkg/controller/pxc/users.go:229]: why don't we declare this in return types in L226?

Comment (qJkee) [reply pkg/controller/pxc/users.go]: fixed in 7d97b53

File: deploy/cr.yaml (hunk 1 above)
Comment (qJkee) [reply deploy/cr.yaml]: reverted

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| scaling | passed |
| haproxy | passed |
| storage | passed |
| scheduled-backup | passed |
| upgrade-haproxy | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| init-deploy | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| limits | passed |
| self-healing-chaos | passed |
| monitoring-2-0 | passed |
| self-healing-advanced | passed |
| affinity | passed |
| smart-update | passed |
| one-pod | passed |
| self-healing-advanced-chaos | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/981/commits/269422bc3c4d4d0e643afb79e074ed5643c99a15
image: `perconalab/percona-xtradb-cluster-operator:PR-981-269422bc`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scheduled-backup | passed |
| scaling | passed |
| upgrade-haproxy | passed |
| haproxy | passed |
| big-data | passed |
| recreate | passed |
| storage | passed |
| self-healing | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| self-healing-chaos | passed |
| limits | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| self-healing-advanced | passed |
| smart-update | passed |
| affinity | passed |
| one-pod | passed |
| self-healing-advanced-chaos | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| pitr | passed |
| users | failed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| upgrade-consistency | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-self | failed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/981/commits/7d97b5300e3b4fc60c5e353d4b7a7404cc396fd6
image: `perconalab/percona-xtradb-cluster-operator:PR-981-7d97b530`

File: pkg/controller/pxc/users.go [hunk 3]
```diff
@@ -57,66 +57,66 @@ func (r *ReconcilePerconaXtraDBCluster) reconcileUsers(cr *api.PerconaXtraDBClus
...
 		if cr.CompareVersionWith("1.7.0") >= 0 {
 			// xtrabackup user need more grants for work in version more then 1.7.0
 			err = r.manageXtrabackupUser(cr, &internalSysSecretObj)
 			if err != nil {
-				return nil, nil, errors.Wrap(err, "manage xtrabackup user")
+				return nil, nil, replicaPWUpdated, errors.Wrap(err, "manage xtrabackup user")
 			}
 		}
 		if cr.CompareVersionWith("1.9.0") >= 0 {
 			err = r.manageReplicationUser(cr, &sysUsersSecretObj, &internalSysSecretObj)
```
[the remark this answers is not in the document]
Comment (qJkee) [reply pkg/controller/pxc/users.go:97]: let's leave it as this for now 3f4e0fa
We will refactor it completely if we need more info about password changes in future

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| upgrade-haproxy | passed |
| scaling | passed |
| scheduled-backup | passed |
| storage | passed |
| big-data | passed |
| recreate | passed |
| haproxy | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| security-context | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| self-healing-chaos | passed |
| demand-backup | passed |
| limits | passed |
| self-healing-advanced | passed |
| smart-update | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| self-healing-advanced-chaos | passed |
| auto-tuning | passed |
| pitr | passed |
| proxysql-sidecar-res-limits | passed |
| operator-self-healing | passed |
| users | passed |
| upgrade-consistency | passed |
| operator-self-healing-chaos | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/981/commits/39ffe24406084ed213e0936b5b80d5b47696b85f
image: `perconalab/percona-xtradb-cluster-operator:PR-981-39ffe244`

**Review by delgod — CHANGES_REQUESTED** [review]

File: deploy/cr.yaml [hunk 4]
```diff
@@ -36,10 +36,10 @@ spec:
...
     image: percona/percona-xtradb-cluster:8.0.23-14.1
     autoRecovery: true
-#    expose:
-#      enabled: true
-#      type: LoadBalancer
-#      trafficPolicy: Local
+#     expose:
+#       enabled: true
+#       type: LoadBalancer
+#       trafficPolicy: Local
```
[the remark this answers is not in the document]
Comment (qJkee) [reply deploy/cr.yaml:42]: needed =)

[the remark this answers is not in the document]
Comment (hors) [reply deploy/cr.yaml:42]: @qJkee please see:
If I uncommented these lines and applied  CR:
```
> kubectl apply -f ./deploy/cr.yaml
error: error parsing ./deploy/cr.yaml: error converting YAML to JSON: yaml: line 39: mapping values are not allowed in this context

diff --git a/deploy/cr.yaml b/deploy/cr.yaml
index f754a509..afff5c9b 100644
--- a/deploy/cr.yaml
+++ b/deploy/cr.yaml
@@ -36,9 +36,9 @@ spec:
     size: 3
     image: percona/percona-xtradb-cluster:8.0.23-14.1
     autoRecovery: true
-#     expose:
-#       enabled: true
-#       type: LoadBalancer
+     expose:
+       enabled: true
+       type: LoadBalancer
 #       trafficPolicy: Local
 #      loadBalancerSourceRanges:
 #        - 10.0.0.0/8
```

but if I remove unneeded spaces everything  is ok:
```
> kubectl apply -f ./deploy/cr.yaml
perconaxtradbcluster.pxc.percona.com/cluster1 created
```

**Review by hors — CHANGES_REQUESTED** [review]

File: deploy/cr.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (qJkee) [reply deploy/cr.yaml]: strange, coz for me it works well
I'll revet this changes then
043982e2

File: pkg/pxc/queries/queries.go [hunk 5]
```diff
@@ -92,6 +92,29 @@ func (p *Database) CurrentReplicationChannels() ([]string, error) {
...
+	tx, err := p.db.Begin()
+	if err != nil {
+		return errors.Wrap(err, "start transaction for updating channel password")
+	}
+	_, err = tx.Exec(`STOP REPLICA IO_THREAD FOR CHANNEL ?`, channel)
+	if err != nil {
+		tx.Rollback()
+		return errors.Wrapf(err, "stop replication IO thread for channel %s", channel)
+	}
+	_, err = tx.Exec(`CHANGE MASTER TO MASTER_PASSWORD=? FOR CHANNEL ?`, password, channel)
```
[the remark this answers is not in the document]
Comment (qJkee) [reply pkg/pxc/queries/queries.go]: b5e3af79

**Review by delgod — DISMISSED** [review]

**Review by hors — DISMISSED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| scheduled-backup | passed |
| big-data | passed |
| haproxy | passed |
| recreate | passed |
| scaling | passed |
| upgrade-haproxy | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| self-healing-chaos | passed |
| demand-backup | passed |
| self-healing-advanced | passed |
| smart-update | passed |
| limits | passed |
| monitoring-2-0 | passed |
| self-healing-advanced-chaos | passed |
| affinity | passed |
| one-pod | passed |
| pitr | passed |
| auto-tuning | passed |
| operator-self-healing | passed |
| proxysql-sidecar-res-limits | passed |
| operator-self-healing-chaos | passed |
| upgrade-consistency | passed |
| users | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/981/commits/043982e255d2b7d3feeca179f934f85909875ef2
image: `perconalab/percona-xtradb-cluster-operator:PR-981-043982e2`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| upgrade-haproxy | passed |
| big-data | passed |
| scheduled-backup | failed |
| scaling | passed |
| storage | passed |
| haproxy | passed |
| recreate | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| init-deploy | passed |
| security-context | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| limits | passed |
| self-healing-chaos | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| self-healing-advanced | passed |
| smart-update | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| self-healing-advanced-chaos | passed |
| proxysql-sidecar-res-limits | passed |
| pitr | passed |
| users | passed |
| operator-self-healing | passed |
| upgrade-consistency | passed |
| operator-self-healing-chaos | failed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| demand-backup-encrypted-with-tls | failed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/981/commits/b5e3af797730379afd1bfb0bfdac83daf44b3470
image: `perconalab/percona-xtradb-cluster-operator:PR-981-b5e3af79`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| upgrade-haproxy | passed |
| scaling-proxysql | passed |
| security-context | passed |
| storage | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| self-healing-advanced | passed |
| self-healing-advanced-chaos | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| recreate | passed |
| haproxy | passed |
| big-data | passed |
| restore-to-encrypted-cluster | passed |
| init-deploy | passed |
| demand-backup | passed |
| limits | passed |
| pitr | passed |
| monitoring-2-0 | passed |
| demand-backup-encrypted-with-tls | passed |
| scheduled-backup | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/981/commits/b5e3af797730379afd1bfb0bfdac83daf44b3470
image: `perconalab/percona-xtradb-cluster-operator:PR-981-b5e3af79`

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| upgrade-haproxy | passed |
| haproxy | passed |
| recreate | passed |
| scaling | passed |
| scheduled-backup | passed |
| big-data | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| init-deploy | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| self-healing-chaos | passed |
| limits | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| self-healing-advanced | passed |
| affinity | passed |
| one-pod | passed |
| smart-update | passed |
| self-healing-advanced-chaos | passed |
| auto-tuning | passed |
| pitr | passed |
| operator-self-healing | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| operator-self-healing-chaos | passed |
| upgrade-consistency | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/981/commits/7dfeb4d285400eaa5849049e988eff904bdab8ae
image: `perconalab/percona-xtradb-cluster-operator:PR-981-7dfeb4d2`

**Review by delgod — APPROVED** [review]

**Pull request merged**